src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -166,8 +166,8 @@ internal void Free(SyntaxListPool pool)
         internal CompilationUnitSyntax ParseCompilationUnit()
         {
             return ParseWithStackGuard(
-                ParseCompilationUnitCore,
-                () => SyntaxFactory.CompilationUnit(
+                static @this => @this.ParseCompilationUnitCore(),
```

Comment:
💡 The other option here is you could create an instance member `_parseCompilationUnitCore` to avoid having the extra stack frame

Reply:
It's an interesting tradeoff. I think your suggestion means we allocate a delegate per LanguageParser instance, but we may create many of those over time, FWIW. This way we have a single static delegate, but also an extra stack frame. I anticipate that `ParseCompilationUnitCore` will be inlined well by the JIT. I did also consider rewriting to make `ParseCompilationUnitCore` use a `LanguageParser @this` parameter but it didn't feel right.

Reply:
i actually prefer sam's suggestion (unless we can prove the inlining happens), as preventing stack frames is def desirable.   The allocation per parser doesn't bother me at all.  There are orders of magnitude less parsers than there are recursive calls in the parser where we want to pass a delegate.  

The big downside of sam's suggestion though is simply that it means the code for the delegate has to move far away from the use site, and we have to manually make all these fields and keep them up to date.  So i'm ok keeping like this, and only revising if we have a stack-frame issue in the future.

Reply:
I thought this suggestion was only for ParseCompilationUnit, which I thought wouldn't happen that many times for an instance of the parser. Did I miss something?

I wonder if it would be interesting to have a generator like this:

```cs
partial class C
{
    [CacheDelegate]
    public void MyMethod() { ... }

    public void UseDelegate(Action a) { ... }
    public void M()
    {
        UseDelegate(MyMethod_Delegate);
        UseDelegate(MyMethod); // warning: don't use a method group conversion on a method with CacheDelegateAttribute
    }
}

partial class C
{
    public Action MyMethod_Delegate => field ??= MyMethod;
}
```

Reply:
> I thought this suggestion was only for ParseCompilationUnit, 

Ah fair.  I may have read more into this suggestion.  For ParseCompilationUnit, i don't think it's worth caching teh delegate explicitly.  If, however, we use this pattern in more places (esp. recursive ones) it may be worth it :)